File: src/main/java/de/gematik/test/fhir/ComparisonDiff.java
package de.gematik.test.fhir;

import lombok.val;
import java.util.Collections;
import java.util.EnumMap;
import java.util.List;
import java.util.Map;
import java.util.Objects;

public final class ComparisonDiff {

  public enum Category {
    TYPE_MISMATCH,
    VALUE_MISMATCH,
    MISSING_FIELD,
    UNEXPECTED_FIELD,
    ARRAY_LENGTH_MISMATCH
  }

  public record DiffEntry(
      String path,
      Category category,
      String expected,
      String actual,
      String parentContext
  ) {
    public DiffEntry {
      Objects.requireNonNull(path, "path must not be null");
      Objects.requireNonNull(category, "category must not be null");
    }
  }

  private final List<DiffEntry> entries;

  public ComparisonDiff(List<DiffEntry> entries) {
    this.entries = List.copyOf(entries);
  }

  public List<DiffEntry> entries() {
    return Collections.unmodifiableList(entries);
  }

  public boolean isEqual() {
    return entries.isEmpty();
  }

  public String summary() {
    if (isEqual()) {
      return "No differences found.";
    }

    val counts = new EnumMap<Category, Integer>(Category.class);
    for (val entry : entries) {
      counts.merge(entry.category(), 1, Integer::sum);
    }

    val builder = new StringBuilder();
    builder.append("Differences found: ").append(entries.size());
    for (val category : Category.values()) {
      val value = counts.get(category);
      if (value != null && value > 0) {
        builder.append(" | ").append(category.name()).append('=').append(value);
      }
    }
    return builder.toString();
  }
}

File: src/main/kotlin/de/gematik/test/fhir/ComparisonDiff.kt
package de.gematik.test.fhir

import java.util.EnumMap

class ComparisonDiff(entries: List<DiffEntry>) {
    val entries: List<DiffEntry> = entries.toList()

    enum class Category {
        TYPE_MISMATCH,
        VALUE_MISMATCH,
        MISSING_FIELD,
        UNEXPECTED_FIELD,
        ARRAY_LENGTH_MISMATCH
    }

    data class DiffEntry(
        val path: String,
        val category: Category,
        val expected: String?,
        val actual: String?,
        val parentContext: String?
    ) {
        fun path(): String = this.path
        fun category(): Category = this.category
        fun expected(): String? = this.expected
        fun actual(): String? = this.actual
        fun parentContext(): String? = this.parentContext
    }

    fun entries(): List<DiffEntry> = entries

    fun isEqual(): Boolean = entries.isEmpty()

    fun summary(): String {
        if (isEqual()) {
            return "No differences found."
        }

        val counts = EnumMap<Category, Int>(Category::class.java)
        entries.forEach { entry ->
            counts.merge(entry.category, 1, Int::plus)
        }

        return buildString {
            append("Differences found: ")
            append(entries.size)
            for (category in Category.entries) {
                val value = counts[category]
                if (value != null && value > 0) {
                    append(" | ")
                    append(category.name)
                    append('=')
                    append(value)
                }
            }

            appendLine()
            append("Examples:")
            entries
                .take(MAX_SUMMARY_EXAMPLES)
                .forEach { entry ->
                    appendLine()
                    append(" - ")
                    append(entry.path)
                    append(" [")
                    append(entry.category)
                    append("] expected=")
                    append(entry.expected)
                    append(" actual=")
                    append(entry.actual)
                }

            if (entries.size > MAX_SUMMARY_EXAMPLES) {
                appendLine()
                append(" - ... and ")
                append(entries.size - MAX_SUMMARY_EXAMPLES)
                append(" more")
            }
        }
    }

    private companion object {
        private const val MAX_SUMMARY_EXAMPLES = 5
    }
}

